### ISSUE TITLE
[LSV] Load Store Vectorizer failed to combine two related loads.

### ISSUE THREAD

**Issue by cdevadas** [llvm/llvm-project#97715]
The LSV pass doesn't combine the loads even though the base address remain the same for them.
There are many instances found in the AMDGPU codegen lit test folder. For example, llvm/test/CodeGen/AMDGPU/GlobalISel/fp-atomics-gfx940.ll (in function local_atomic_fadd_v2bf16_noret) should have the two loads combined earlier. But they are merged using the target specific Load Store Optimizer pass (si-load-store-opt) after ISel.

**Comment by jayfoad** [comment]
> llvm/test/CodeGen/AMDGPU/GlobalISel/fp-atomics-gfx940.ll (in function local_atomic_fadd_v2bf16_noret)

I don't see that function in that file. Can you provide a link?

**Comment by cdevadas** [comment]
Looks like I have a slightly older version of upstream compiler. This test has been changed. However the same test compiled for SelectionDAG still exists and it reproduces the problem.
https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/AMDGPU/fp-atomics-gfx940.ll#L364

**Comment by arsenm** [comment]
I think this is just from unhandled insertion of bitcasts to get the types to match for vectorization: 

```
; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx940 -print-after=load-store-vectorizer < %s

define amdgpu_kernel void @no_vectorize_0(ptr addrspace(3) %ptr, <2 x half> %data) {
  %i1 = atomicrmw fadd ptr addrspace(3) %ptr, <2 x half> %data syncscope("agent") seq_cst, align 4
  ret void
}

define amdgpu_kernel void @no_vectorize_1(i32 %ptr.as.int, <2 x half> %data) {
  %ptr = inttoptr i32 %ptr.as.int to ptr addrspace(3)
  %i1 = atomicrmw fadd ptr addrspace(3) %ptr, <2 x half> %data syncscope("agent") seq_cst, align 4
  ret void
}

define amdgpu_kernel void @does_vectorize(i32 %ptr.as.int, i32 %data.as.int) {
  %ptr = inttoptr i32 %ptr.as.int to ptr addrspace(3)
  %data = bitcast i32 %data.as.int to <2 x half>
  %i1 = atomicrmw fadd ptr addrspace(3) %ptr, <2 x half> %data syncscope("agent") seq_cst, align 4
  ret void
}

```

**Comment by gandhi56** [comment]
This issue appears to be resolved for the above test and `@flat_atomic_fadd_v2bf16_noret` from llvm/test/CodeGen/AMDGPU/GlobalISel/fp-atomics-gfx940.ll. Should we close this issue, @arsenm @cdevadas ?

**Comment by arsenm** [comment]
> This issue appears to be resolved for the above test and `@flat_atomic_fadd_v2bf16_noret` from llvm/test/CodeGen/AMDGPU/GlobalISel/fp-atomics-gfx940.ll. Should we close this issue, @arsenm @cdevadas ?

No, this is not fixed at all. The specific function instance doesn't matter. You can see this is any kernel signature with mixed types. See the output of my last comment. The first 2 functions are more representative cases, and the last kernel is the manually coerced version. The vectorizer fails to pack the first 2, and the MIR pass has to fix up the missed vectorization

**Comment by gandhi56** [comment]
It seems that vectorization only occurs when the type of `%data` is i32 after removing all bitcasts.

**Comment by arsenm** [comment]
> It seems that vectorization only occurs when the type of `%data` is i32 after removing all bitcasts.

Yes, it's easiest to handle a bunch of uniformly typed arguments. The weird mismatched vectors and scalars aren't handled

**Comment by gandhi56** [comment]
```
define amdgpu_kernel void @no_vectorize_0(ptr addrspace(3) %ptr, <2 x half> %data) {
  %no_vectorize_0.kernarg.segment = call nonnull align 16 dereferenceable(264) ptr addrspace(4) @llvm.amdgcn.kernarg.segment.ptr()

  ; load pointer
  %ptr.load = load ptr addrspace(3), ptr addrspace(4) %no_vectorize_0.kernarg.segment, align 16

  ; load data
  %data.kernarg.offset = getelementptr inbounds i8, ptr addrspace(4) %no_vectorize_0.kernarg.segment, i64 4
  %data.load = load <2 x half>, ptr addrspace(4) %data.kernarg.offset, align 4

  %i1 = atomicrmw fadd ptr addrspace(3) %ptr.load, <2 x half> %data.load syncscope("agent") seq_cst, align 4
  ret void
}
```

In this example, `%ptr.load` can be replaced by a `load i32` and an `inttoptr`, and `%data.load` can be replaced by a `load i32` and a `bitcast i32 to <2 x half>`. This way, LSV should be able to vectorize the two loads into a `load <2 x i32>`. This seems scalable for all allowed types for `atomicrmw fadd`.

**Comment by gandhi56** [comment]
Created a pull request: https://github.com/llvm/llvm-project/pull/114702

**Comment by arsenm** [comment]
> Created a pull request: #114702

The atomicrmw has nothing to do with the issue here. It is simply a sample user instruction. The problem is the source load instructions with mismatched types

**Comment by gandhi56** [comment]
Okay, so we need a solution that inserts bitcasts and inttoptr's to maximize load/store vectorization within each pseudo-BB in question, right? What do you mean by "source" load instructions?

**Comment by arsenm** [comment]
> Okay, so we need a solution that inserts bitcasts and inttoptr's to maximize load/store vectorization within each pseudo-BB in question, right? What do you mean by "source" load instructions?

I mean the loads in the example question providing the values for the atomicrmw, not the atomicrmw itself.

The inserted casts to coerce the type need to be directly on the loads or stores being vectorized

**Comment by gandhi56** [comment]
```
define amdgpu_kernel void @does_vectorize(i32 %ptr.as.int, i32 %data.as.int) {
  %ptr = inttoptr i32 %ptr.as.int to ptr addrspace(3)
  %data = bitcast i32 %data.as.int to <2 x half>
  %i1 = atomicrmw fadd ptr addrspace(3) %ptr, <2 x half> %data syncscope("agent") seq_cst, align 4
  ret void
}
```
I don't think i32 can be directly casted to <2 x half>. The vectorizer would need to insert sitofp, fptrunc and 2 insertelement instructions to ensure that the value is represented correctly. So that's adding about 6 instructions to vectorize a <2 x half> with other i32's, is that really profitable?

**Comment by gandhi56** [comment]
Update: I have a branch with my implementation so far: https://github.com/llvm/llvm-project/compare/main...gandhi56:llvm-project:issue-97715/classify-based-on-typesize

A side-effect of this patch, as illustrated in some of the tests as FIXMEs, is that multiple vector instructions may be generated which can be merged. To solve that, I am thinking of merging equivalence classes before the actual vectorization.

**Comment by gandhi56** [comment]
I created a PR: https://github.com/llvm/llvm-project/pull/134436. I think I could be more selective about choosing the type to cast the elements into. I will work on that.

**Comment by dakersnar** [comment]
Happy to discuss this more. FYI I'm away on Thursday and Friday, so will be more active starting next week.

**Comment by Artem-B** [comment]
+cc: @jlebar who implemented the current version of the algorithm.

**Comment by jlebar** [comment]
My 2c: If I was trying to work on this today, I would immediately write a fuzzer.

Essentially:
 - create a random program P
 - let P' be the program after running LSV
 - simulate P and P' to check that they effectively do the same sequence of loads and stores

You'd have to write a simulator for a (very small) subset of LLVM IR, but I don't expect this would be particularly hard.

I'm sure you'd find a zillion bugs.  Also when we have a new idea like this one, we don't have to worry that it's going to silently introduce correctness issues.

**Comment by gandhi56** [comment]
> Hypothetical 1: Unpacking and bit casting are "free", performance wise. Solution: The vectorizer should not care at all about the element type of the final vector or the underlying chain elements, until the very end. It should define ChainElems based purely on offset and bytes accessed, attempt to create a vector load of a sequence of bytes that is the superset of as many chain elements as possible, then extract/bitcast the resulting vector back to the types that the original loads needed.

Agreed, we could probably keep track of the original type in `ChainElem` and perform all decision making in-memory until vectorization. Casts can finally be inserted in `vectorizeChain`.

> Hypothetical 2: Unpacking and bit casting are _not_ 100% free. Solution: The vectorizer should follow some heuristic to try to choose the final granularity of the vector based on the "majority" of types of the chain elements, to minimize bitcasts. So for instance, a chain of {i32, i32, i16, i8, i8, i32} could result in a <4 x i32>, bitcasting only for the three elements that do not match this type size.

We could take the gcd over bitwidths of the chain elements to determine the normalized type, with a limit in place regarding the number of bitcasts we are inserting - although I don't think we need to worry about them. We can always be aggressive about the casts DCE them if they aren't used.

> Hypothetical 3: Some targets have a preference on the shape of a vector output by the LSV. For example, preferring a <4 x i32> over a <16 x i8>. NVPTX _does not_ have this preference, as we have designed our backend to "upsize" vectors of the "wrong" shape during lowering, but that may not be true for all targets. Solution: The vectorizer should use some TTI call to help make the decision on the final granularity of the vector.

AFAIK, AMDGPU does not have a preference either. 

> (P.S. just like your original motivator, this is a problem we have seen in our NVPTX compiler pipeline. For example, when unpacking loads/stores to aggregates in InstCombine, if padding of structs is represented as i8s, we have issues with later vectorization. For instance, if a struct is {i32, i32, i8, i8, i8, i8, i32}. Our current internal approach is to merge the i8s into an i32 in InstCombine to enable later Vectorization, but there should be a less hacky way of doing this.)

**Comment by dakersnar** [comment]
All sounds good to me. Would you like to attempt a prototype of this? And I can collaborate in the PR to help iterate on it. Some more small details:

>  we could probably keep track of the original type in ChainElem and perform all decision making in-memory until vectorization. Casts can finally be inserted in vectorizeChain.

It's encoded on the Inst, so we may not need to add a new field to ChainElem. What could _maybe_ be worth caching is the result of getTypeStoreSize, which would let us avoid having to call that multiple times throughout the vectorizer. But if it isn't that expensive of an access, we could just derive it from the Inst.

> We could take the gcd over bitwidths of the chain elements to determine the normalized type, with a limit in place regarding the number of bitcasts we are inserting - although I don't think we need to worry about them. We can always be aggressive about the casts DCE them if they aren't used.

Yeah I'm sure we could come up with a clean heuristic. Let's maybe try to compartmentalize some of this vectorizeChain logic while we are at it. Selecting the vector shape and emitting the right bitcasts could each be made into helper functions to keep the logic clean.

Also, I want to call this out now: it's possible the "linear"/single-pass nature of the current vectorizer algorithm may need to be reworked to accommodate this change. I'm trying to imagine a specific edge case, but there could be cases where the vectorizer selects a sub-optimal vector because the "candidate chains" in any of the split functions are not guaranteed to be sorted best-to-worst anymore, if that makes sense. I think we should be able to design around this, but it would be good to keep this in mind.

**Comment by gandhi56** [comment]
Sounds good to me. I will start working on a prototype tonight or tomorrow.

**Comment by dakersnar** [comment]
I'm out sick for at least today. I'll hopefully be available for review within the next week.
